### PR TITLE
[interna/sharedcomponent] Use a ring buffer to restrict remembered status count

### PR DESCRIPTION
**Pull request by TylerHelmuth**
#### Description
Use a ring buffer to only remember the last 5 events. This ensures we remember a reasonable number of events during startup, so that a status aggregator gets the events for all instances. Then during normal operation, when we're done adding sources and no longer need to replay events, we don't have to remember every single event.

#### Link to tracking issue
Closes https://github.com/open-telemetry/opentelemetry-collector/issues/11818

#### Testing

`go test status_test.go -count 1000 -failfast` still passes with many tries.

### CODE REVIEW

File: internal/sharedcomponent/sharedcomponent.go
Comment (bogdandrutu) [inline internal/sharedcomponent/sharedcomponent.go:122]: How can this happen?

Comment (TylerHelmuth) [reply internal/sharedcomponent/sharedcomponent.go:122]: Added a test.  It can happen if a component were to ever create an event and then emit it twice.  Components should really never do that, but this defensive code makes sure that if they do we don't fill up our buffer with duplicate events.

Comment (bogdandrutu) [reply internal/sharedcomponent/sharedcomponent.go:122]: Another thing is that events have a "timestamp" right now witch makes them to be unique every time you create a new event, but there is a discussion to remove that, which will cause lots of "duplicate" events to be detected and that is wrong most likely.

So overall I think this is a "premature defensive code" (optimization) that we should remove.

Comment (TylerHelmuth) [reply internal/sharedcomponent/sharedcomponent.go:122]: ok

**Review by bogdandrutu — APPROVED** [review]

**Comment by mx-psi** [comment]
@TylerHelmuth Ready to merge once you fix the merge conflicts

**Review by codeboten — APPROVED** [review]

**Pull request merged**
